### ISSUE TITLE
API Section 6.4: Cloning versus groups of Connections?

### ISSUE THREAD

**Issue by gorryfair** [taps-api/drafts#64]
I'm curious about whether cloning is the correct way to go:

Maybe (I don't know) the concept of cloning could fit well with postsockets and less well with NEAT.? - NEAT does have the concept of combining flows into a group - and relative priorities within the group, etc. So is this actually the way we need to go? (and then they can share attributes)
- I'm worried about how cloning before a race works after racing the candidates, and after it has found candidates, I am not sure all the parameters can then be copied. 

I see you can likely do some of these things if the connections are within an SCTP association, or a QUIC session, but the present description seems very linked to these transports. 

Basically can cloning  be implemented? And haow is racing handled?

**Comment by britram** [comment]
Could you elaborate a bit your concerns about implementability? The implementation of clone() seems pretty straightforward to me:

* An established Connection has a set of properties (retained by reference or by value from its Preconnection, perhaps with certain of the properties and preferences collapsed). 

* If the Connection's selected candidate Protocol Stack is multistreaming, and additional streams are available, a new stream is created. If it is not multistreaming, then a new connection is created (perhaps using the post-race properties, maybe by starting over from the retained Preconnection -- that really seems like an implementation choice)

* When a new stream is created/opened/first used by the remote endpoint, a new Connection is cloned off the/a Connection associated with another stream; this is experienced by the "passively cloning" application as a Ready<> event.

Okay, cloning a connection that is still racing seems like it could be difficult to get right. The easy implementation is to basically buffer any sends (within the API) on these cloned connections until the "master" (i.e. first opened) connection is established well enough to understand whether multistreaming is available. It's not clear to me that the less easy implementations will perform better in any other than pathological cases.

**Comment by britram** [comment]
(From a finishing-the-edit-pass standpoint, I'm inclined to stay with cloning for London; we can revisit this discussion for -01)

**Comment by gorryfair** [comment]
So... your different perspective may be that you see "cloning" as some way to abstract multistreaming. Hence my comment "I see you can likely do some of these things if the connections are within an SCTP association, or a QUIC session, but the present description seems very linked to these transports."

But you are also able to group multiple connections without multistreaming. Sharing the way they encode DCSPs in a PHB group, TCB-like congestion state, flow labels, etc - this doesn't imply a multi-streaming solution - but it does give relative priority and some notion of shared transport function.  That's more what I was thinking, and is nearer what we decided to implement.

Unsure what others think on this.

**Comment by britram** [comment]
So it seems natural to me that a clone of a connection maintains some relationship with its antecedents, and therefore naturally forms a group, whether that group be bound in the same underlying protocol stack or not. If we want to add connection groups to the `clone()` concept that seems perfectly reasonable to me.

The "connection group" concept as in NEAT, with an id as a reference (at least as I understand it from Michael's comments on it) seems very `setsockopt()`-like to me, but I can't really explain my aesthetic aversion to it in detail at the moment.

**Comment by tfpauly** [comment]
I also share the concern that using group IDs as the primary entrance to multiplexing gets pretty sockopt-like. One concern could be that if I'm in an application that's using a global networking context, I'm not sure who's maintaining this group ID namespace, and two pieces of code within the same process could overlap.

Cloning makes it very explicit that you can multiplex explicitly with another Connection when you already have a handle to the other/original Connection.

Either way, I think can leave Clone for now, and continue the discussion, since this is all very interesting =)

**Comment by mwelzl** [comment]
My view is: whether the functionality of grouping is exposed as a group ID or via "clone" is just syntactical sugar. I don't see this or that way as better, and it's all the same to me - irrespective of whether we're talking about multiple streams of an association or connections that share the same DSCP or whatever.  One important thing that shouldn't be lost is that they are "entangled", i.e. changing parameters on one changes the other - because that's inevitable in the mux case, and so in any other cases of grouping it must be implemented below the API for consistency.

**Comment by philsbln** [comment]
I guess we will need two kinds of connection groups: 
- one that represents multi-streaming connections
- one that represents connection pools between the same endpoints

Also resulting in two different clone calls. Not sure how to avoid them to be mixed up easily.

**Comment by mwelzl** [comment]
Whaaat???  Noooo!   Why???

**Comment by philsbln** [comment]
When you want to do per-message path selection, you can do this in two ways:
- Make different connections over different paths an assign messages to the connections
- Make one multi-path multi-streaming connection and assign messages to path.

Both should be supported by a transport system without making life too miserable for the application.

In addition, connection pools are a so widely used pattern in modern transport systems that we should also include them. I guess most applications that call ``clone`` just want "something" between the same endpoints (that does not suffer hidden head-of-line blocking), not necessarily multi-streams of the same connection.

But that is definitely a discussion we should have in London.

**Comment by philsbln** [comment]
Small clarification: I see connection pooling not as an replacement for multi-streaming – both are orthogonal. Whether a connection pool represents a bunch of open TCP connections between two endpoints or a single QUIC connection does not matter to the application.
The difference ist that a connection pool can hold connections between the same endpoints, but with different parameters, our connection group can't.

**Comment by abrunstrom** [comment]
The current text has a problem in that the entanglement has been extended to the closing of the Connection. "closing one of them also closes all others". Closing is not the same as changing parameters and should not be tied together. This is also inconsistent with what we have in the implementation draft and needs to be removed.

**Comment by mwelzl** [comment]
True, sorry, that was a bug. Fixed!

**Comment by britram** [comment]
Decided at side meeting in London: remove `Clone()` from `Preconnection`, rename "Multiplex Connections" to "Multistream Connections in Group"
